docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,400 @@
+curve used, points in `G1` are represented with 48 bytes and points in `G2` are
+represented with 96 bytes. It is up to the implementer of the cryptosystem to
+decide which should be larger, the public keys or the signatures.
+
+BLS signatures rely on pairing-based elliptic-curve cryptography to produce
+various types of signatures. For a more in-depth but still high level discussion
+pairing-based elliptic-curve cryptography, see Vitalik Buterin's post on
+[Exploring Elliptic Curve Pairings][vitalik-curve-post]. For much more in
+depth discussion, see the specific paper on BLS12-381, 
+[Compact Multi-Signatures for Smaller Blockchains][multi-signatures-smaller-blockchains].
+
+### Adoption
+
+BLS signatures have already gained traction within several popular projects.
+
+* Algorand is working on an implementation.
+* [Zcash][zcash-adoption] has adopted BLS12-381 into the protocol.
+* [Ethereum 2.0][eth-2-adoption] has adopted BLS12-381 into the protocol.
+* [Chia Network][chia-adoption] has adopted BLS for signing blocks.
+* [Ostracon][line-ostracon-pr], a fork of Tendermint has adopted BLS for signing blocks.
+
+### What systems may be affected by adding aggregated signatures?
+
+#### Gossip
+
+Gossip could be updated to aggregate vote signatures during a consensus round.
+This appears to be of frankly little utility. Creating an aggregated signature
+incurs overhead, so frequently re-aggregating may incur a significant
+overhead. How costly this is is still subject to further investigation and
+performance testing.
+
+Each validator will still need to receive vote extension data
+from the peer validators in order for consensus to proceed. As a result, any
+advantage gained by aggregating signatures across the vote message will be
+nullified as a result of the addition of vote extensions.
+
+#### Block Creation
+
+When creating a block, the proposer may create a small set of short
+multi-signatures and attach these to the block instead of including one
+signature per validator.
+
+#### Block Verification
+
+Verification of blocks would not verify a set of many signatures. Verification
+would instead check the single multi-signature using the public keys stored
+by the validator. Currently, we verify each validator signature using
+the public key associated with that validator.
+
+#### IBC Relaying
+
+IBC would no longer need to transmit a large set of signatures when
+updating state. These state updates do not happen for every IBC packet, only
+when changing an IBC light client's view of the counterparty chain's state.
+General [packets][ibc-packet] only contain enough information to correctly
+route the data to the counterparty chain.
+
+IBC does persist commit signatures to the chain in these `MsgUpdateClient`
+message when updating state. This message would no longer need the full set
+of unique signatures and would instead only need one signature for all of the
+data in the header.
```

Comment:
```suggestion
IBC does persist commit signatures to the chain in a `MsgUpdateClient`
message when updating state. This message would no longer need the full set
of unique signatures, but would instead only need one aggregated signature for
all of the data in the header.
```

Not sure: Is aggregated sufficient here, or is an identifying multi-signature needed?

Reply:
That's a good question. I _believe_ you need to know _which_ validators signed the block, but I'm not certain of that offhand. After an initial review by our team, I'd like to expand this out to the IBC team to provide additional feedback and advice where possible so hopefully this can be addressed then.

Reply:
My thought was along similar lines—I expect you'd need to know the identities of the validators, since how else would you match votes to voting power to check 2/3+? Given that how much VP a validator has can change over time.

Reply:
Yes, in my sketch of an implementation, I believe this would look like including that `BitArray` of which validators were included in the signature so that IBC could check which validator votes were contained in the signature.

Reply:
That makes sense. In which case I think we _do_ need to be able to identify signers.

Reply:
To clarify, the bit array would allow us to identify signers, if I'm understanding you correctly. Basically, you need the bit array to identify which subset, `S`, of the validators included signatures in the aggregated signature. You then perform an aggregate verification with the public keys of all members of `S`.

Reply:
Got it. So we really _do_ need to keep the individual keys, so that we can choose which ones to use for verifying the aggregate signature.

Reply:
Yes, that's correct.

Reply:
In the `MsgUpdateClient` the relayers include a `Header` field, which in the case of a normal light client update is of [type `Header`](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/07-tendermint/types/tendermint.pb.go#L187-L204). And this type includes the current validator set with the voting power of each validator, which is used in the light client verification algorithm.

Reply:
So if I understand correctly, `MsgUpdateClient` would need to additionally take in a bit array to select which validators from the passed in validator set is a signer of the aggregate signature?

> And this type includes the current validator set with the voting power of each validator, which is used in the light client verification algorithm.

This is correct. The full validator set will need to be passed in though to pass the [valSet hash check](https://github.com/tendermint/tendermint/blob/cbae5f9f5336bd9882cb0b02912f7ae948fac5b4/light/verifier.go#L266), so the bit array would still be necessary